### ISSUE TITLE
训练数据生成 

### ISSUE THREAD

**Issue by lrain-CN** [goodluckcwl/Face-alignment-mobilenet-v2#5]
想问下 网络输入是64*64，是不是需要修改训练数据的landmark坐标为检测框扣出来的人脸图，然后resize到64*64的相对坐标。

另外ImageFloatData输入的txt是 image_path x1,y1,x2,y2 ...x68,y68 这样的格式嘛。
谢谢解惑~

**Comment by goodluckcwl** [comment]
你好，是相对坐标，需要归一化。ImageFloatData输入格式是x1,x2,...,x68,y1,y2,...,y68
------------------
------------------
Best regards,

Willem,Chen
Graduate
College of Information Science & Electronic Engineering
Zhejiang University,Hangzhou,P.R.C.

 
陈伟亮
浙江大学信电学院 2016级信通硕士

Tel:(+86)18868107517
Email: willemchen@foxmail.com












------------------ 原始邮件 ------------------
发件人: "lrain-CN"<notifications@github.com>;
发送时间: 2018年9月17日(星期一) 下午3:54
收件人: "goodluckcwl/Face-alignment-mobilenet-v2"<Face-alignment-mobilenet-v2@noreply.github.com>;
抄送: "Subscribed"<subscribed@noreply.github.com>; 
主题: [goodluckcwl/Face-alignment-mobilenet-v2] 训练数据生成  (#5)




想问下 网络输入是6464，是不是需要修改训练数据的landmark坐标为检测框扣出来的人脸图，然后resize到6464的相对坐标。
 
另外ImageFloatData输入的txt是 image_path x1,y1,x2,y2 ...x68,y68 这样的格式嘛。
 谢谢解惑~

**Comment by lrain-CN** [comment]
感谢回复~  试了下发现代码preprocess函数中resize后的图像shape是63*63。应该是不需要round 和 -1的，直接  bbox[2:] = bbox[0:2] + [input_width, input_height]   bbox = bbox.astype(np.int32) 是可以的。
还有最后的landmark归一化操作不是很理解含义，是否是必须的 还是说对训练有帮助~

**Comment by goodluckcwl** [comment]
归一化一般是需要的


------------------
------------------
Best regards,

Willem,Chen
Graduate
College of Information Science & Electronic Engineering
Zhejiang University,Hangzhou,P.R.C.

 
陈伟亮
浙江大学信电学院 2016级信通硕士

Tel:(+86)18868107517
Email: willemchen@foxmail.com






 





------------------ 原始邮件 ------------------
发件人: "lrain-CN"<notifications@github.com>;
发送时间: 2018年9月18日(星期二) 晚上7:32
收件人: "goodluckcwl/Face-alignment-mobilenet-v2"<Face-alignment-mobilenet-v2@noreply.github.com>;
抄送: "陈伟亮"<willemchen@foxmail.com>; "Comment"<comment@noreply.github.com>; 
主题: Re: [goodluckcwl/Face-alignment-mobilenet-v2] 训练数据生成  (#5)




感谢回复~  试了下发现代码preprocess函数中resize后的图像shape是63*63。应该是不需要round 和 -1的，直接  bbox[2:] = bbox[0:2] + [input_width, input_height]   bbox = bbox.astype(np.int32) 是可以的。
 还有最后的landmark归一化操作不是很理解含义，是否是必须的 还是说对训练有帮助~

**Comment by songsong1014** [comment]
@goodluckcwl 您好 想问一下您的这个ImageFloatData输入的txt完整格式是什么呢？看您的回复的格式，发现txt中没有image_path，不是很清楚想问一下，谢谢。

**Comment by LingQiu** [comment]
您好，我在用您修改的mobilenet-v2做fine-tune，数据是按照您这里的提示制作的，caffe是用您修改的custom-caffe，训练过程中遇到了些问题，下面是部分日志
`I1109 13:43:30.693302 10676 net.cpp:242] This network produces output landmarkloss
I1109 13:43:30.708902 10676 net.cpp:255] Network initialization done.
I1109 13:43:30.708902 10676 solver.cpp:56] Solver scaffolding done.
I1109 13:43:30.724503 10676 caffe.cpp:156] Finetuning from mobilenet-v2.caffemodel
I1109 13:43:31.800902 10676 upgrade_proto.cpp:77] Attempting to upgrade batch norm layers using deprecated params: mobilenet-v2.caffemodel
I1109 13:43:31.800902 10676 upgrade_proto.cpp:80] Successfully upgraded batch norm layers using deprecated params.
I1109 13:43:31.800902 10676 net.cpp:745] Ignoring source layer noise
I1109 13:43:31.800902 10676 caffe.cpp:253] Starting Optimization
I1109 13:43:31.800902 10676 solver.cpp:278] Solving 
I1109 13:43:31.800902 10676 solver.cpp:279] Learning Rate Policy: inv
I1109 13:43:31.816503 10676 blocking_queue.cpp:49] Waiting for data
I1109 13:43:37.276502 10676 solver.cpp:224] Iteration 0 (0 iter/s, 5.47079s/200 iters), loss = inf
I1109 13:43:37.276502 10676 solver.cpp:243]     Train net output #0: landmarkloss = inf (* 1 = inf loss)
I1109 13:43:37.276502 10676 sgd_solver.cpp:137] Iteration 0, lr = 1e-06
I1109 13:43:37.292102 10676 sgd_solver.cpp:169] scale layer:0.069746 0.024842 0.046855 0.060976 0.009700 0.016846 0.089128 0.015259 0.021536 0.055264 0.010337 0.017989 0.060302 0.030748 0.048405 0.285392 0.008099 0.013739 0.108765 
I1109 13:43:37.292102 10676 sgd_solver.cpp:192] weight diff/data:nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan inf 
I1109 13:43:42.206102 10676 solver.cpp:214] Iteration 1, loss = nan, ignore and continue.
I1109 13:43:47.510103 10676 solver.cpp:214] Iteration 2, loss = nan, ignore and continue.
I1109 13:43:52.845304 10676 solver.cpp:214] Iteration 3, loss = nan, ignore and continue.
I1109 13:43:57.986304 10676 solver.cpp:214] Iteration 4, loss = nan, ignore and continue.
I1109 13:44:03.617903 10676 solver.cpp:214] Iteration 5, loss = nan, ignore and continue.
I1109 13:44:08.859503 10676 solver.cpp:214] Iteration 6, loss = nan, ignore and continue.
I1109 13:44:14.116703 10676 solver.cpp:214] Iteration 7, loss = nan, ignore and continue.
I1109 13:44:19.436303 10676 solver.cpp:214] Iteration 8, loss = nan, ignore and continue.
I1109 13:44:24.646703 10676 solver.cpp:214] Iteration 9, loss = nan, ignore and continue.
I1109 13:44:29.763504 10676 solver.cpp:214] Iteration 10, loss = nan, ignore and continue.
I1109 13:44:34.833503 10676 solver.cpp:214] Iteration 11, loss = nan, ignore and continue.
I1109 13:44:39.794303 10676 solver.cpp:214] Iteration 12, loss = nan, ignore and continue.
I1109 13:44:44.567903 10676 solver.cpp:214] Iteration 13, loss = nan, ignore and continue.
I1109 13:44:49.762703 10676 solver.cpp:214] Iteration 14, loss = nan, ignore and continue.
I1109 13:44:54.879503 10676 solver.cpp:214] Iteration 15, loss = nan, ignore and continue.
I1109 13:45:00.167903 10676 solver.cpp:214] Iteration 16, loss = nan, ignore and continue.
I1109 13:45:05.549903 10676 solver.cpp:214] Iteration 17, loss = nan, ignore and continue.
I1109 13:45:10.756903 10676 solver.cpp:214] Iteration 18, loss = nan, ignore and continue.
I1109 13:45:12.782304 10676 solver.cpp:214] Iteration 19, loss = nan, ignore and continue.
I1109 13:45:13.266903 10676 solver.cpp:214] Iteration 20, loss = nan, ignore and continue.
I1109 13:45:13.772104 10676 solver.cpp:214] Iteration 21, loss = nan, ignore and continue.
I1109 13:45:14.265703 10676 solver.cpp:214] Iteration 22, loss = nan, ignore and continue.
I1109 13:45:14.749303 10676 solver.cpp:214] Iteration 23, loss = nan, ignore and continue.
I1109 13:45:15.238903 10676 solver.cpp:214] Iteration 24, loss = nan, ignore and continue.
I1109 13:45:15.722503 10676 solver.cpp:214] Iteration 25, loss = nan, ignore and continue.
I1109 13:45:16.206104 10676 solver.cpp:214] Iteration 26, loss = nan, ignore and continue.
I1109 13:45:16.689703 10676 solver.cpp:214] Iteration 27, loss = nan, ignore and continue.
I1109 13:45:17.188904 10676 solver.cpp:214] Iteration 28, loss = nan, ignore and continue.
I1109 13:45:17.672503 10676 solver.cpp:214] Iteration 29, loss = nan, ignore and continue.
I1109 13:45:18.156103 10676 solver.cpp:214] Iteration 30, loss = nan, ignore and continue.
I1109 13:45:18.639703 10676 solver.cpp:214] Iteration 31, loss = nan, ignore and continue.
I1109 13:45:19.123303 10676 solver.cpp:214] Iteration 32, loss = nan, ignore and continue.
I1109 13:45:19.622503 10676 solver.cpp:214] Iteration 33, loss = nan, ignore and continue.
I1109 13:45:20.107103 10676 solver.cpp:214] Iteration 34, loss = nan, ignore and continue.
I1109 13:45:20.590703 10676 solver.cpp:214] Iteration 35, loss = nan, ignore and continue.
I1109 13:45:21.089903 10676 solver.cpp:214] Iteration 36, loss = nan, ignore and continue.
I1109 13:45:21.573503 10676 solver.cpp:214] Iteration 37, loss = nan, ignore and continue.
I1109 13:45:22.057103 10676 solver.cpp:214] Iteration 38, loss = nan, ignore and continue.
I1109 13:45:22.556303 10676 solver.cpp:214] Iteration 39, loss = nan, ignore and continue.
I1109 13:45:23.055503 10676 solver.cpp:214] Iteration 40, loss = nan, ignore and continue.
I1109 13:45:23.539103 10676 solver.cpp:214] Iteration 41, loss = nan, ignore and continue.
I1109 13:45:23.991503 10676 solver.cpp:214] Iteration 42, loss = nan, ignore and continue.
I1109 13:45:24.490703 10676 solver.cpp:214] Iteration 43, loss = nan, ignore and continue.
I1109 13:45:24.989903 10676 solver.cpp:214] Iteration 44, loss = nan, ignore and continue.
I1109 13:45:25.489104 10676 solver.cpp:214] Iteration 45, loss = nan, ignore and continue.
I1109 13:45:25.975703 10676 solver.cpp:214] Iteration 46, loss = nan, ignore and continue.
I1109 13:45:26.475903 10676 solver.cpp:214] Iteration 47, loss = nan, ignore and continue.
I1109 13:45:26.943903 10676 solver.cpp:214] Iteration 48, loss = nan, ignore and continue.
I1109 13:45:27.427503 10676 solver.cpp:214] Iteration 49, loss = nan, ignore and continue.
I1109 13:45:27.926703 10676 solver.cpp:214] Iteration 50, loss = nan, ignore and continue.`

不明白为什么loss为nan，是我做的数据有问题吗？还有就是在solver.prototxt已经设置迭代200次显示一下，这里为什么每次都显示，通过调试发现确实有这部分输出，是不是solver.prototxt里面的设置没有用了？

**Comment by goodluckcwl** [comment]
每次都显示是因为loss是nan。loss为什么是nan,我觉得你可以检查下输入数据对不对

**Comment by LingQiu** [comment]
感谢回复。以下是部分预处理的效果图：
![default](https://user-images.githubusercontent.com/20528352/48254822-fe973980-e445-11e8-8d24-fdc6608de19a.png)
图像预处理以及归一化代码如下：
` # 预处理图像
 bbox = np.array(bbox, dtype=np.float32)
 bbox = obtain_bbox(bbox, input_width * 1.0 / input_height)
 bbox = pad_bbox(bbox, pad_ratio=0.15)

 cropImg, pad, offset, scale = preprocess(img, bbox)
            
 landmarks_x = (np.array(landmark_x) + 1 + pad) / scale - offset[0]
 landmarks_y = (np.array(landmark_y) + 1 + pad) / scale - offset[1]
  # 归一化
  landmarks_normal_x = (landmarks_x + 1) / input_width
  landmarks_normal_y = (landmarks_y + 1) / input_height
`
`bbox`中的数据来自标注文件的68个点的最小、最大坐标，不知道这种处理方式可以吗？目前loss依然为nan。

**Comment by goodluckcwl** [comment]
是这样的，一开始训练的时候，先用L2损失函数（normalized:0），训练到收敛了再用瞳距加权的L2损失函数(normalized:1)。
layer {
  name: "landmarkloss"
  type: "WeightEuclideanLoss"
  bottom: "fc4"
  bottom: "landmark"
  top: "landmarkloss"
  loss_weight: 1
  include {
    phase: TRAIN
  }
  weight_euclidean_loss_param {
    contour_weight: 1
    main_weight: 1
    coarse_weight: 1
 normalized: 0

  }
}

**Comment by WallZFE** [comment]
您好，我在这里看到ImageFloatData输入的TXT完整格式是image_dir，x1，x2，... x68，y1，y2，...，y68，但是您的caffe上label_dim是136，则该怎么解决，谢谢

**Comment by lrain-CN** [comment]
> 
> 
> 您好，我在这里看到ImageFloatData输入的TXT完整格式是image_dir，x1，x2，... x68，y1，y2，...，y68，但是您的caffe上label_dim是136，则该怎么解决，谢谢

68*2就是136啊

**Comment by WallZFE** [comment]
> > 你好，我在这里看到ImageFloatData输入的TXT完整格式是image_dir，x1，x2，... x68，y1，y2，...，y68，但是你的caffe上标签_dim是136，则该怎么解决，谢谢
> 
> 68 * 2就是136啊

主要是前面还写了图片的名字，这样不就是有137了吗

**Comment by lrain-CN** [comment]
> 
> 
> > > 你好，我在这里看到ImageFloatData输入的TXT完整格式是image_dir，x1，x2，... x68，y1，y2，...，y68，但是你的caffe上标签_dim是136，则该怎么解决，谢谢
> > 
> > 
> > 68 * 2就是136啊
> 
> 主要是前面还写了图片的名字，这样不就是有137了吗

图片路径是用来读图片数据的，和label是没关系。 感觉你这是caffe的prototxt 还是caffe编译的有问题

> 
> 
> > > 你好，我在这里看到ImageFloatData输入的TXT完整格式是image_dir，x1，x2，... x68，y1，y2，...，y68，但是你的caffe上标签_dim是136，则该怎么解决，谢谢
> > 
> > 
> > 68 * 2就是136啊
> 
> 谢谢您回复我，谢谢谢谢。主要是前面还写了图片的名字，这样不就是有137了吗
> 这是我的txt文件
> indoor_001.png 446.000 449.459 450.957 460.552 471.486 488.087 506.016 524.662 553.315 578.732 599.361 615.947 627.439 635.375 642.063 647.302 646.518 470.271 486.218 503.097 519.714 533.680 571.937 585.702 602.344 620.077 633.964 554.931 554.589 554.658 554.777 532.717 543.637 555.652 565.441 576.368 487.474 499.184 513.781 527.594 513.422 499.117 579.876 590.901 605.648 618.343 606.771 593.466 519.040 536.292 547.001 557.161 568.172 579.144 589.098 581.071 570.103 558.241 547.661 534.085 524.758 547.684 557.304 568.172 585.417 569.211 557.473 547.989 91.000 119.344 150.614 176.986 202.157 226.842 246.438 263.865 271.435 266.260 248.966 220.651 197.999 179.064 156.371 124.753 92.944 117.870 109.415 114.454 120.090 127.609 123.590 117.155 109.070 103.951 111.236 145.072 161.106 177.570 194.295 197.930 202.841 205.483 202.069 197.061 136.436 132.337 133.589 143.047 144.769 144.737 140.815 130.008 128.376 132.671 140.525 141.419 229.040 221.978 221.192 224.381 219.826 222.233 224.410 239.804 251.962 254.844 254.621 247.772 230.477 231.663 230.805 229.159 225.992 237.777 240.542 240.014
> 在运行时的错误
> I0313 03:41:38.479493 72 layer_factory.hpp:77] Creating layer data
> I0313 03:41:38.479523 72 net.cpp:84] Creating Layer data
> I0313 03:41:38.479532 72 net.cpp:380] data -> data
> I0313 03:41:38.479557 72 net.cpp:380] data -> landmark
> I0313 03:41:38.479614 72 image_float_data_layer.cpp:40] Opening file /home/data_2/train_name.txt
> F0313 03:41:38.479698 72 image_float_data_layer.cpp:67] Check failed: label_dim_ == label.size() (136 vs. 137)


图片路径是用来读图片数据的，和label是没关系。 感觉你这是caffe的prototxt 还是caffe编译的有问题

**Comment by lrain-CN** [comment]
> 
> 
> > > 你好，我在这里看到ImageFloatData输入的TXT完整格式是image_dir，x1，x2，... x68，y1，y2，...，y68，但是你的caffe上标签_dim是136，则该怎么解决，谢谢
> > 
> > 
> > 68 * 2就是136啊
> 
> 主要是前面还写了图片的名字，这样不就是有137了吗

图片路径是用来读图片数据的，和label是没关系。 感觉你这是caffe的prototxt 还是caffe编译的有问题
